io-sim/src/Control/Monad/IOSim/Internal.hs

```diff
@@ -809,15 +982,23 @@ unblockThreads !wakeup !simstate@SimState {runqueue, threads} =
 unwindControlStack :: forall s a.
                       SomeException
                    -> Thread s a
-                   -> Either Bool (Thread s a)
+                                               --v If the ControlStack had a
+                                               --  TimeoutFrame
+                   -> ( Either Bool (Thread s a, Maybe TimeoutId)
+                      --v If ControlStack had a ThreadDelayFrame
+                      , Maybe TimeoutId
+                      )
 unwindControlStack e thread =
     case threadControl thread of
       ThreadControl _ ctl -> unwind (threadMasking thread) ctl
   where
     unwind :: forall s' c. MaskingState
-           -> ControlStack s' c a -> Either Bool (Thread s' a)
-    unwind _  MainFrame                 = Left True
-    unwind _  ForkFrame                 = Left False
+           -> ControlStack s' c a
+           -> ( Either Bool (Thread s' a, Maybe TimeoutId)
+              , Maybe TimeoutId
+              )
```

Comment:
I'm not convinced that we need to be collecting and returning these timeout ids. I think that's due to the need to delete the locks right? If we follow the locking design I suggested then the lifetime of locks becomes simpler.

Or was it due to being strict about deleting thread delay timeouts?

Reply:
Yes, we can simplify this!